[BUG] Zenflow_stage3 - RuntimeError: narrow() cannot be applied to a 0-dim tensor.

**Describe the bug**
When running ZeRO Stage 3 with the ZenFlow optimizer, the training fails during the backward pass on the first optimization step (`model_engine.backward(loss)`).

The crash is caused by a `RuntimeError: narrow() cannot be applied to a 0-dim tensor` originating inside the `group_step` function of the class `ZenFlowSelectiveAdamW_stage3`.

Detailed debugging revealed a corruption in the internal DeepSpeed parameter state: The parameter's metadata (`param.ds_shape`) correctly reports a 2D tensor (e.g., `torch.Size([3072, 1024])`), but the parameter's underlying data handle (`param.ds_tensor.data`) is incorrectly initialized as a 0-dimensional scalar (`torch.Size([])`), leading to the crash when the narrow() method is called on it.

**To Reproduce**
Steps to reproduce the behavior:
1. Configure a DeepSpeed JSON file (`ds_config_zf_shell.json`) to enable ZeRO Stage 3 with Optimizer Offload and the ZenFlow extension.
2. Launch the finetune_zf.py script using the DeepSpeed launcher.

**ds_config_zf_shell.json**
```json
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": 2e-5,
            "betas": [
                0.9,
                0.95
            ],
            "eps": 1e-8,
            "weight_decay": 0.01
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": 0.0,
            "warmup_max_lr": 2e-5,
            "warmup_num_steps": 100
        }
    },
    "zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device": "nvme",
            "pin_memory": true,
            "nvme_path": "/mnt/nvme1",
            "buffer_count": 5
        },
        "offload_param": {
            "device": "nvme",
            "pin_memory": true,
            "nvme_path": "/mnt/nvme1",
            "buffer_count": 5,
            "buffer_size": 5e9,
            "max_in_cpu": 1e8
        },
        "zenflow": {
            "topk_ratio": 0.05,
            "update_interval": 4,
            "full_warm_up_rounds": 0,
            "overlap_step": true
        },
        "overlap_comm": true,
        "contiguous_gradients": true,
        "sub_group_size": 1e6,
        "reduce_bucket_size": 2e8,
        "stage3_prefetch_bucket_size": 2e7,
        "stage3_param_persistence_threshold": 1e6,
        "stage3_max_live_parameters": 1e8,
        "stage3_max_reuse_distance": 0,
        "stage3_gather_16bit_weights_on_model_save": true
    },
    "activation_checkpointing": {
        "partition_activations": true,
        "cpu_checkpointing": false,
        "contiguous_memory_optimization": true,
        "number_checkpoints": null,
        "synchronize_checkpoint_boundary": false
    },
    "comms_logger": {
        "enabled": true,
        "verbose": false,
        "prof_all": true,
        "debug": false
    },
    "gradient_accumulation_steps": 1,
    "gradient_clipping": 1.0,
    "steps_per_print": 2000,
    "train_batch_size": 4,
    "train_micro_batch_size_per_gpu": 4,
    "wall_clock_breakdown": false,
    "zero_allow_untested_optimizer": true
}
```

**finetune_zf.py**
```python
import torch
import time
import deepspeed
import argparse
import os
import random
import numpy as np
import datetime
from datasets import load_dataset
from datasets.formatting.formatting import LazyBatch
from torch.utils.data import DataLoader
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    DataCollatorForLanguageModeling,
)
from deepspeed import comm as dist

os.environ["TOKENIZERS_PARALLELISM"] = "false"

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)

def format_chat_template(question: str, answer: str, tokenizer) -> str:
    messages = [
        {'role': 'user', 'content': question},
        {'role': 'assistant', 'content': answer},
    ]
    
    chat_template = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=False,
    )
    return chat_template

def tokenize_function(examples: LazyBatch, tokenizer, max_length: int) -> torch.Tensor:
    chat_templates = [
        format_chat_template(question, answer, tokenizer)
        for question, answer in zip(examples['prompt'], examples['response'])
    ]
    
    return tokenizer(
        chat_templates, 
        truncation=True, 
        padding='max_length', 
        max_length=max_length, 
        return_tensors='pt'
    )


def log_event(message: str):
    if dist.get_rank() == 0:
        # Get current time with microseconds
        now = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')
        print(f"[{now}] {message}", flush=True) # flush=True ensures it prints immediately

def main(args):    
    # Initialize DeepSpeed
    # This will automatically handle setting the device (e.g., args.local_rank)
    deepspeed.init_distributed()
    
    set_seed(args.seed)

    # Load Tokenizer
    tokenizer = AutoTokenizer.from_pretrained(args.model_path)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
        
    if dist.get_rank() == 0:
        print("Tokenizer loaded.")

    # Load Model
    model = AutoModelForCausalLM.from_pretrained(
        args.model_path,
    )
    
    # Resize token embeddings if tokenizer has new tokens
    model.resize_token_embeddings(len(tokenizer))
    
    # --- End of Fix ---
    if dist.get_rank() == 0:
        print(f"Model loaded. Embedding size: {model.get_input_embeddings().weight.shape}")

    # Load and Process Dataset
    dataset = load_dataset('Dahoas/rm-static', split='train')
    # Use a smaller subset for quick testing if needed
    # samples = dataset.select(range(3000))
    
    tokenized_dataset = dataset.map(
        tokenize_function,
        batched=True,
        # Pass tokenizer and max_length to the map function
        fn_kwargs={'tokenizer': tokenizer, 'max_length': args.max_seq_len},
        remove_columns=dataset.column_names # Keep only tokenized columns
    )
    
    if dist.get_rank() == 0:
        print("Dataset processed.")

    # Data Collator
    data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

    # DeepSpeed Initialization
    model_engine, optimizer, train_dataloader, lr_scheduler = deepspeed.initialize(
        args=args,
        model=model,
        model_parameters=model.parameters(),
        training_data=tokenized_dataset,
        collate_fn=data_collator
    )
    
    if dist.get_rank() == 0:
        print("DeepSpeed engine initialized.")

    # --- Training Loop ---
    model_engine.train()
    global_step = 0
    
    if dist.get_rank() == 0:
        print("Starting training...")

    for epoch in range(args.num_train_epochs):
        if dist.get_rank() == 0:
            print(f"--- Starting Epoch {epoch + 1}/{args.num_train_epochs} ---")
            
        for step, batch in enumerate(train_dataloader):
            log_event(f"[Step {global_step + 1}][start]")
            step_start_time = time.time()
            
            # Move batch to the correct device
            batch = {k: v.to(model_engine.device) for k, v in batch.items()}
            
            # Forward pass
            log_event("[Forward][start]")
            outputs = model_engine(**batch)
            loss = outputs.loss

            # --- 2. Backward Pass ---
            log_event("[Backward][start]")
            # Backward pass
            model_engine.backward(loss)

            log_event("[Update][start]")
            # Optimizer step
            model_engine.step()

            step_time = time.time() - step_start_time
            global_step += 1
            
            # Log to console
            if dist.get_rank() == 0 and global_step % args.log_steps == 0:
                print(f"Step {global_step:5d} | Loss: {loss.item():.4f} | Time: {step_time*1000:.0f}ms | LR: {lr_scheduler.get_last_lr()[0]:.2e}")
            log_event(f"[Step {global_step}][end]")
            # Check if max_steps is reached
            if args.max_steps > 0 and global_step >= args.max_steps:
                if dist.get_rank() == 0:
                    print(f"Reached max_steps ({args.max_steps}). Stopping training.")
                break
        
        if args.max_steps > 0 and global_step >= args.max_steps:
            break

        log_event(f"[Epoch {epoch + 1}][end]")


    # --- End of Training ---

    # Save checkpoint
    if dist.get_rank() == 0:
        print("Training finished. Saving checkpoint...")
        model_engine.save_checkpoint(args.output_dir)
        # Also save the tokenizer for easy-to-use-later
        tokenizer.save_pretrained(args.output_dir)
        print(f"Model and tokenizer saved to {args.output_dir}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="DeepSpeed Core Training Script (No Trainer)")
    
    # --- Script-specific arguments ---
    parser.add_argument("--model_path", type=str, required=True,
                        help="Path to the pretrained model (e.g., /usr/local/models/Llama-3.2-3B-Instruct)")
    parser.add_argument("--output_dir", type=str, required=True,
                        help="Directory to save checkpoints and tokenizer")
    parser.add_argument("--max_seq_len", type=int, default=1024,
                        help="Maximum sequence length for tokenization")
    parser.add_argument("--num_train_epochs", type=int, default=1,
                        help="Number of training epochs")
    parser.add_argument("--max_steps", type=int, default=-1,
                        help="Total number of training steps. Overrides num_train_epochs. (-1 to ignore)")
    parser.add_argument("--log_steps", type=int, default=1,
                        help="Log training status every X steps")
    parser.add_argument("--seed", type=int, default=42,
                        help="Random seed for reproducibility")
    
    # --- DeepSpeed arguments ---
    # This argument is required by deepspeed.initialize()
    # It tells DeepSpeed to find its config (e.g., --deepspeed ds_config.json)
    parser.add_argument('--local_rank',
                        type=int,
                        default=-1,
                        help='local rank passed from distributed launcher')
    
    # Add all of DeepSpeed's arguments
    parser = deepspeed.add_config_arguments(parser)
    
    args = parser.parse_args()

    main(args)
```

**finetune_zf.sh**
```bash
#!/bin/bash
export CUDA_DEVICE_MAX_CONNECTIONS=1
export CUDA_VISIBLE_DEVICES=7

GPUS_PER_NODE=1
NNODES=1
NODE_RANK=0
WORLD_SIZE=$(($GPUS_PER_NODE * $NNODES))

# Model parameters
MODEL_PATH="/usr/local/models/Llama-3.2-3B-Instruct"
OUTPUT_DIR="/User/deepspeed/output/deepspeed"
EPOCHS=3
SEED=42
MAX_SEQ_LEN=512 # Added this parameter

# ZenFlow config file path
DS_CONFIG_JSON="./ds_config_zf_shell.json"

# Note: LR, batch_size, weight_decay are defined in the config file
# These parameters are kept for fallback only
LR=2e-5
BATCH_SIZE=8
WARMUP=0.03
WEIGHT_DECAY=0.01

# Create output directory if it doesn't exist
mkdir -p $OUTPUT_DIR

# DeepSpeed command
if [ -f "$DS_CONFIG_JSON" ]; then
    echo "[INFO] Using DeepSpeed config file: $DS_CONFIG_JSON"
    echo "[INFO] Forcing execution on CUDA_VISIBLE_DEVICES=$CUDA_VISIBLE_DEVICES"
    CMD="deepspeed finetune_zf.py \
        --deepspeed_config=$DS_CONFIG_JSON \
        --model_path $MODEL_PATH \
        --num_train_epochs $EPOCHS \
        --output_dir $OUTPUT_DIR \
        --seed $SEED \
        --max_seq_len $MAX_SEQ_LEN"

else
    echo "[ERROR] DeepSpeed config file not found: $DS_CONFIG_JSON"
    exit 1
fi

echo "[INFO] Running DeepSpeed training with ZenFlow:"
echo $CMD
eval $CMD
```

**Expected behavior**
-

**ds_report output**
```
(deepspeed) user@userserver:~/deepspeed$ uv run ds_report
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
dc ..................... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
 [WARNING]  FP Quantizer is using an untested triton version (3.5.0), only 2.3.(0, 1) and 3.0.0 are known to be compatible with these kernels
fp_quantizer ........... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
gds .................... [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.9
 [WARNING]  using untested triton version (3.5.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/User/deepspeed/.venv/lib/python3.10/site-packages/torch']
torch version .................... 2.9.0+cu128
deepspeed install path ........... ['/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.18.2, unknown, unknown
torch cuda version ............... 12.8
torch hip version ................ None
nvcc version ..................... 12.8
deepspeed wheel compiled w. ...... torch 0.0, cuda 0.0
shared memory (/dev/shm) size .... 503.78 GB
```

**Screenshots**
If applicable, add screenshots to help explain your problem.
```bash
(deepspeed) user@userserver:~/deepspeed$ uv run ./finetune_zf.sh
[INFO] Using DeepSpeed config file: ./ds_config_zf_shell.json
[INFO] Forcing execution on CUDA_VISIBLE_DEVICES=7
[INFO] Running DeepSpeed training with ZenFlow:
deepspeed finetune_zf.py --deepspeed_config=./ds_config_zf_shell.json --model_path /usr/local/models/Llama-3.2-3B-Instruct --num_train_epochs 3 --output_dir /User/deepspeed/output/deepspeed --seed 42 --max_seq_len 512
[2025-11-12 08:27:30,890] [WARNING] [runner.py:232:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.
Detected VISIBLE_DEVICES=7: setting --include=localhost:7
[2025-11-12 08:27:30,890] [INFO] [runner.py:630:main] cmd = /User/deepspeed/.venv/bin/python3 -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbN119 --master_addr=127.0.0.1 --master_port=29500 --enable_each_rank_log=None --log_level=info finetune_zf.py --deepspeed_config=./ds_config_zf_shell.json --model_path /usr/local/models/Llama-3.2-3B-Instruct --num_train_epochs 3 --output_dir /User/deepspeed/output/deepspeed --seed 42 --max_seq_len 512
[2025-11-12 08:27:36,225] [INFO] [launch.py:162:main] WORLD INFO DICT: {'localhost': [7]}
[2025-11-12 08:27:36,225] [INFO] [launch.py:168:main] nnodes=1, num_local_procs=1, node_rank=0
[2025-11-12 08:27:36,225] [INFO] [launch.py:179:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0]})
[2025-11-12 08:27:36,225] [INFO] [launch.py:180:main] dist_world_size=1
[2025-11-12 08:27:36,225] [INFO] [launch.py:184:main] Setting CUDA_VISIBLE_DEVICES=7
[2025-11-12 08:27:36,226] [INFO] [launch.py:272:main] process 2334180 spawned with command: ['/User/deepspeed/.venv/bin/python3', '-u', 'finetune_zf.py', '--local_rank=0', '--deepspeed_config=./ds_config_zf_shell.json', '--model_path', '/usr/local/models/Llama-3.2-3B-Instruct', '--num_train_epochs', '3', '--output_dir', '/User/deepspeed/output/deepspeed', '--seed', '42', '--max_seq_len', '512']
Tokenizer loaded.
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.64it/s]
Model loaded. Embedding size: torch.Size([128256, 3072])
Dataset processed.
ZenFlowCPUAdam initialized with overlap step.
[2025-11-12 08:27:54,912] [WARNING] [engine.py:1406:_do_optimizer_sanity_check] **** You are using ZeRO with an untested optimizer, proceed with caution *****
Parameter Offload - Persistent parameters statistics: param_count = 57, numel = 175104
[2025-11-12 08:29:15,712] [WARNING] [lr_schedules.py:690:get_lr] Attempting to get learning rate from scheduler before it has started
DeepSpeed engine initialized.
Starting training...
--- Starting Epoch 1/3 ---
[2025-11-12 08:29:15.837165] [Step 1][start]
[2025-11-12 08:29:15.837882] [Forward][start]
[2025-11-12 08:29:18.439815] [Backward][start]
Zenflow: clearing selective optimizer states...
[rank0]: Traceback (most recent call last):
[rank0]:   File "/User/deepspeed/finetune_zf.py", line 239, in <module>
[rank0]:     main(args)
[rank0]:   File "/User/deepspeed/finetune_zf.py", line 172, in main
[rank0]:     model_engine.backward(loss)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 20, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 2356, in backward
[rank0]:     self._do_optimizer_backward(loss, retain_graph)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 2297, in _do_optimizer_backward
[rank0]:     self.optimizer.backward(loss, retain_graph=retain_graph)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 20, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zero/stage3.py", line 2341, in backward
[rank0]:     self.loss_scaler.backward(loss.float(), retain_graph=retain_graph)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/fp16/loss_scaler.py", line 65, in backward
[rank0]:     scaled_loss.backward(retain_graph=retain_graph)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/torch/_tensor.py", line 625, in backward
[rank0]:     torch.autograd.backward(
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/torch/autograd/__init__.py", line 354, in backward
[rank0]:     _engine_run_backward(
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/torch/autograd/graph.py", line 841, in _engine_run_backward
[rank0]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 20, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zero/stage3.py", line 1247, in reduce_partition_and_remove_grads
[rank0]:     self.reduce_ready_partitions_and_remove_grads(param)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zero/stage3.py", line 1573, in reduce_ready_partitions_and_remove_grads
[rank0]:     self.reduce_independent_p_g_buckets_and_remove_grads(param)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zero/stage3.py", line 1294, in reduce_independent_p_g_buckets_and_remove_grads
[rank0]:     self.__reduce_and_partition_ipg_grads(comm_dtype)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 20, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zero/stage3.py", line 1362, in __reduce_and_partition_ipg_grads
[rank0]:     self._process_selected_fp32_groups_grad(params_in_bucket, grad_partitions)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/runtime/zenflow/engine_stage3.py", line 327, in _process_selected_fp32_groups_grad
[rank0]:     optimizer_z3.selective_optimizer.group_step(params_to_update)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/User/deepspeed/.venv/lib/python3.10/site-packages/deepspeed/ops/adam/zenflow_torch_adam.py", line 425, in group_step
[rank0]:     param_2d = param.ds_tensor.data.narrow(0, param.complete_column_offset,
[rank0]: RuntimeError: narrow() cannot be applied to a 0-dim tensor.
[2025-11-12 08:29:42,354] [INFO] [launch.py:335:sigkill_handler] Killing subprocess 2334180
[2025-11-12 08:29:42,355] [ERROR] [launch.py:341:sigkill_handler] ['/User/deepspeed/.venv/bin/python3', '-u', 'finetune_zf.py', '--local_rank=0', '--deepspeed_config=./ds_config_zf_shell.json', '--model_path', '/usr/local/models/Llama-3.2-3B-Instruct', '--num_train_epochs', '3', '--output_dir', '/User/deepspeed/output/deepspeed', '--seed', '42', '--max_seq_len', '512'] exits with return code = 1
```

**System info (please complete the following information):**
 - OS: Ubuntu 22.04
 - GPU count and types: 1 machine with 1 * A6000
 - Python version: 3.10.12
 - Any other relevant info about your setup

**Launcher context**
Launched with `deepspeed` launcher

**Docker context**
-

**Additional context**
-

--- a/deepspeed/ops/adam/zenflow_torch_adam.py
+++ b/deepspeed/ops/adam/zenflow_torch_adam.py
@@ -283,12 +283,14 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
     def temp_copy_param(self, paramlist):
         for param in paramlist:
             if hasattr(param, "selected_grad"):
+                swapped_in = self._swap_in_if_offloaded(param)
                 num_column, num_row = param.ds_shape if len(param.ds_shape) != 1 else (param.ds_shape[0], 1)
 
                 if num_row != 1:
+                    indices = param.selected_indices.to(param.ds_tensor.data.device)
                     param_2d = param.ds_tensor.data.narrow(0, param.complete_column_offset, param.complete_numel).view(
                         param.complete_numel // num_row, num_row)
-                    temp_selected_param = param_2d[param.selected_indices, :].clone().detach()
+                    temp_selected_param = param_2d[indices, :].clone().detach()
                 else:
                     temp_selected_param = param.ds_tensor.data.clone().detach()
 
@@ -296,6 +298,8 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
                     param.temp_selected_param = temp_selected_param.cpu()
                 else:
                     param.temp_selected_param = temp_selected_param
+
+                self._release_if_offloaded(param, swapped_in)
 
     def clear_selected_mv(self):
         print("Zenflow: clearing selective optimizer states...")
@@ -314,6 +318,10 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
     @torch.no_grad()
     def _step_without_offload(self):
         for group in self.param_groups:
+
+            if any(hasattr(p, "selected_grad") and self._is_partition_offloaded(p) for p in group["params"]):
+                self._group_step_offloaded(group["params"], group)
+                continue
 
             params_with_grad: List[Tensor] = []
             grads: List[Tensor] = []
@@ -392,6 +400,129 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
             param.exp_avg = None
             param.exp_avg_sq = None
 
+    def _is_partition_offloaded(self, param):
+        """Whether the parameter partition needs the per-parameter offload path.
+
+        The selective update assumes the partition is resident on the compute device. When
+        parameters are offloaded this no longer holds: an NVMe partition is a 0-dim placeholder,
+        and a CPU partition lives on a different device than the gradients and optimizer state.
+        """
+        data = param.ds_tensor.data
+        return data.dim() == 0 or data.device != param.selected_grad.device
+
+    def _swap_in_if_offloaded(self, param):
+        """Read an NVMe-offloaded partition back into memory.
+
+        The swapper leaves ``ds_tensor.data`` as a 0-dim placeholder while a partition is on NVMe.
+        Returns True when this call performed the swap-in, so the caller owns the matching swap-out.
+        """
+        if param.ds_tensor.data.dim() == 0 and getattr(param, "nvme_swapper", None) is not None:
+            param.nvme_swapper.swap_in([param], async_op=False)
+            return True
+        return False
+
+    def _swap_out_if_offloaded(self, param, swapped_in):
+        """Persist an updated partition back to NVMe and free its compute buffer."""
+        if swapped_in:
+            param.nvme_swapper.swap_out_and_release([param])
+
+    def _release_if_offloaded(self, param, swapped_in):
+        """Free a partition that was swapped in only to be read.
+
+        Its contents are unchanged, so it is not written back to NVMe.
+        """
+        if swapped_in:
+            param.nvme_swapper.remove_partition_and_release_buffers([param])
+
+    def _selective_step_one_param(self, param, group, apply_temp=False):
+        """Run the selective AdamW update for a single parameter.
+
+        The partition (``ds_tensor.data``) may live on CPU when offloaded, while the gradient and
+        optimizer state live on the compute device. The math runs on the compute device and the
+        result is written back to wherever the partition resides.
+        """
+        amsgrad = group["amsgrad"]
+        beta1, beta2 = cast(Tuple[float, float], group["betas"])
+        num_column, num_row = param.ds_shape if len(param.ds_shape) != 1 else (param.ds_shape[0], 1)
+        data = param.ds_tensor.data
+        data_device = data.device
+        compute_device = param.selected_grad.device
+
+        if num_row != 1:
+            indices = param.selected_indices.to(data_device)
+            param_2d = data.narrow(0, param.complete_column_offset,
+                                   param.complete_numel).view(param.complete_numel // num_row, num_row)
+            selected_param = param_2d[indices, :].to(compute_device)
+        else:
+            selected_param = data.to(compute_device)
+
+        if apply_temp and getattr(param, "temp_selected_param", None) is not None:
+            selected_param.copy_(param.temp_selected_param.to(compute_device))
+
+        grad = param.selected_grad.to(compute_device)
+
+        state = self.state.setdefault(param, {})
+        if len(state) == 0:
+            state["step"] = torch.zeros((), dtype=param.dtype, device=compute_device)
+            if amsgrad:
+                state["max_exp_avg_sq"] = torch.zeros_like(selected_param)
+            if not self.offload:
+                state["exp_avg"] = torch.zeros_like(selected_param)
+                state["exp_avg_sq"] = torch.zeros_like(selected_param)
+
+        if self.offload:
+            # Restore the offloaded moment/variance to the compute device and keep the references
+            # so copy_mv_to_cpu writes the updated tensors back.
+            exp_avg_t = param.exp_avg.to(compute_device).view_as(selected_param)
+            exp_avg_sq_t = param.exp_avg_sq.to(compute_device).view_as(selected_param)
+            param.exp_avg = exp_avg_t
+            param.exp_avg_sq = exp_avg_sq_t
+        else:
+            exp_avg_t = state["exp_avg"]
+            exp_avg_sq_t = state["exp_avg_sq"]
+
+        adamw(
+            [selected_param],
+            [grad],
+            [exp_avg_t],
+            [exp_avg_sq_t],
+            [state["max_exp_avg_sq"]] if amsgrad else [],
+            [state["step"]],
+            amsgrad=amsgrad,
+            beta1=beta1,
+            beta2=beta2,
+            lr=group["lr"],
+            weight_decay=group["weight_decay"],
+            eps=group["eps"],
+            maximize=False,
+        )
+
+        if num_row != 1:
+            param_2d[indices, :] = selected_param.to(data_device)
+        else:
+            data.copy_(selected_param.to(data_device))
+
+    @torch.no_grad()
+    def _group_step_offloaded(self, params, group):
+        """Per-parameter selective update for offloaded partitions (CPU or NVMe).
+
+        NVMe partitions are swapped in, updated, and swapped back out one at a time so at most one
+        partition is resident and the swapper buffer pool is never exhausted.
+        """
+        for param in params:
+            if not hasattr(param, "selected_grad"):
+                continue
+            swapped_in = self._swap_in_if_offloaded(param)
+            if self.offload:
+                self.copy_mv_from_cpu([param])
+            self._selective_step_one_param(param, group, apply_temp=True)
+            if self.offload:
+                self.copy_mv_to_cpu([param])
+            self._swap_out_if_offloaded(param, swapped_in)
+            if hasattr(param, "temp_selected_param"):
+                param.temp_selected_param = None
+            param.selected_grad = None
+
     @torch.no_grad()
     def group_step(self, paramlist):
 
@@ -405,6 +536,10 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
         for group_id in sorted(group_to_paramlist.keys()):
             params = group_to_paramlist[group_id]
             group = self.param_groups[group_id]
+
+            if any(hasattr(p, "selected_grad") and self._is_partition_offloaded(p) for p in params):
+                self._group_step_offloaded(params, group)
+                continue
 
             if self.offload:
                 self.copy_mv_from_cpu(params)
@@ -507,6 +642,10 @@ class ZenFlowSelectiveAdamW_stage3(torch.optim.AdamW):
                     return
                 for param in bucket:
                     if hasattr(param, "temp_selected_param") and param.temp_selected_param is not None:
+                        # Offloaded partitions are not resident on the compute device here;
+                        # group_step applies the snapshot per-parameter on the right device.
+                        if self._is_partition_offloaded(param):
+                            continue
                         temp_selected_param = param.temp_selected_param.to(param.device, non_blocking=True)
                         num_column, num_row = param.ds_shape if len(param.ds_shape) != 1 else (param.ds_shape[0], 1)
                         if num_row != 1:

--- a/tests/unit/ops/adam/test_zf_torch_adam.py
+++ b/tests/unit/ops/adam/test_zf_torch_adam.py
@@ -185,3 +185,99 @@ def test_against_torch_adamw(Opt):
     param = make_param(Opt, (2, 4), torch.arange(4))
     opt = Opt([param], lr=1e-3, offload=False)
     _compare_with_torch_adamw(param, opt)
+
+
+class _FakeNvmeSwapper:
+    """Minimal stand-in for the partitioned-param NVMe swapper.
+
+    It mimics the contract the stage3 selective optimizer relies on: an offloaded
+    partition leaves ``ds_tensor.data`` as a 0-dim placeholder, swap-in restores the
+    stored partition, and swap-out persists it and restores the placeholder. This lets
+    us exercise the swap-in/update/swap-out path (issue #7686) without real NVMe.
+    """
+
+    def __init__(self):
+        self.storage = {}
+        self.swap_in_calls = 0
+        self.swap_out_calls = 0
+
+    def _placeholder(self, param):
+        return torch.tensor(1).to(param.ds_tensor.dtype)
+
+    def offload(self, param):
+        self.storage[param.ds_id] = param.ds_tensor.data.clone()
+        param.ds_tensor.data = self._placeholder(param)
+
+    def swap_in(self, params, async_op=False):
+        for param in params:
+            param.ds_tensor.data = self.storage[param.ds_id]
+        self.swap_in_calls += 1
+
+    def swap_out_and_release(self, params, async_op=False):
+        for param in params:
+            self.storage[param.ds_id] = param.ds_tensor.data.clone()
+            param.ds_tensor.data = self._placeholder(param)
+        self.swap_out_calls += 1
+
+    def remove_partition_and_release_buffers(self, params):
+        for param in params:
+            param.ds_tensor.data = self._placeholder(param)
+
+
+def _offloaded_stage3_param(selected_indices):
+    param = make_param(ZenFlowSelectiveAdamW_stage3, (4, 6), selected_indices)
+    param.ds_id = 0
+    swapper = _FakeNvmeSwapper()
+    param.nvme_swapper = swapper
+    swapper.offload(param)
+    assert param.ds_tensor.data.dim() == 0
+    return param, swapper
+
+
+def test_group_step_nvme_offloaded_partition():
+    # Regression test for issue #7686: group_step must not crash when the partition is
+    # offloaded to NVMe (ds_tensor.data is a 0-dim placeholder).
+    param, swapper = _offloaded_stage3_param(torch.tensor([0, 1, 3]))
+    before = swapper.storage[param.ds_id].clone()
+
+    opt = ZenFlowSelectiveAdamW_stage3([param], lr=1e-3, offload=False)
+    opt.group_step([param])
+
+    assert swapper.swap_in_calls == 1
+    assert swapper.swap_out_calls == 1
+    assert param.ds_tensor.data.dim() == 0, "partition should be re-offloaded after the step"
+    assert param.selected_grad is None
+    after = swapper.storage[param.ds_id]
+    assert (before - after).abs().sum().item() > 1e-5, "offloaded partition was not updated"
+
+
+def test_step_nvme_offloaded_partition():
+    # The full step() entry point must also handle an offloaded partition.
+    param, swapper = _offloaded_stage3_param(torch.tensor([0, 2, 4]))
+    before = swapper.storage[param.ds_id].clone()
+
+    opt = ZenFlowSelectiveAdamW_stage3([param], lr=1e-3, offload=False)
+    opt.step()
+
+    assert swapper.swap_in_calls == 1
+    assert swapper.swap_out_calls == 1
+    assert param.ds_tensor.data.dim() == 0
+    assert param.temp_selected_param is None
+    assert param.selected_grad is None
+    after = swapper.storage[param.ds_id]
+    assert (before - after).abs().sum().item() > 1e-5
+
+
+def test_temp_copy_param_nvme_offloaded_partition():
+    # temp_copy_param only reads the partition, so it must release it without writing back.
+    param, swapper = _offloaded_stage3_param(torch.tensor([1, 2, 5]))
+    stored = swapper.storage[param.ds_id].clone()
+
+    opt = ZenFlowSelectiveAdamW_stage3([param], lr=1e-3, offload=False)
+    opt.temp_copy_param([param])
+
+    assert swapper.swap_in_calls == 1
+    assert swapper.swap_out_calls == 0, "read-only snapshot must not write back to NVMe"
+    assert param.ds_tensor.data.dim() == 0
+    assert param.temp_selected_param is not None
+    assert (stored - swapper.storage[param.ds_id]).abs().sum().item() == 0

